### PR TITLE
feat: secret reference graph for understanding how its pulled

### PR DESCRIPTION
**Pull request by akhilmhdh**
# Description 📣

This PR introduces a secret reference details window in detail side bar. This helpers user to see the original value and also a graph view inference made to pull the reference. 

The function is attached to the original function so that both are kept in sync. The next thing is tracing is optional inside the function so that most of other the time expansion happens its not done

Screenshot 2024-10-17 at 4 29 45 PM

## Type ✨

- [ ] Bug fix
- [x] New feature
- [ ] Improvement
- [ ] Breaking change
- [ ] Documentation

# Tests 🛠️

```sh
# Here's some code block to paste some code snippets
```

---

- [ ] I have read the [contributing guide](https://infisical.com/docs/contributing/getting-started/overview), agreed and acknowledged the [code of conduct](https://infisical.com/docs/contributing/getting-started/code-of-conduct). 📝

### CODE REVIEW

**Comment by akhilmhdh** [comment]
Don't merge this until permission v2 is merged

File: backend/src/services/secret-v2-bridge/secret-v2-bridge-service.ts
Comment (scott-ray-wilson) [inline backend/src/services/secret-v2-bridge/secret-v2-bridge-service.ts:1708]: Isn't this needlessly crawling references twice? it's split into two functions in `secret-v2-bridge-fns` but they both run the same function returning a different property. Maybe just making `stackTrace` an optional prop to `expandSecretReferences`?

File: frontend/src/views/SecretMainPage/components/SecretListView/SecretDetailSidebar.tsx
Comment (scott-ray-wilson) [inline frontend/src/views/SecretMainPage/components/SecretListView/SecretDetailSidebar.tsx:434]: nit: `Reference Tree`

File: frontend/src/views/SecretMainPage/components/SecretListView/SecretDetailSidebar.tsx
Comment (scott-ray-wilson) [inline frontend/src/views/SecretMainPage/components/SecretListView/SecretDetailSidebar.tsx:439]: How do you feel about something like `Visual breakdown of secrets referenced by this secret`?

File: frontend/src/views/SecretMainPage/components/SecretListView/SecretDetailSidebar.tsx
Comment (scott-ray-wilson) [inline frontend/src/views/SecretMainPage/components/SecretListView/SecretDetailSidebar.tsx:431]: Thoughts on `faDiagramProject` instead?

File: frontend/src/views/SecretMainPage/components/SecretListView/SecretDetailSidebar.tsx
Comment (scott-ray-wilson) [inline frontend/src/views/SecretMainPage/components/SecretListView/SecretDetailSidebar.tsx:432]: I think we should also check if this is even a secret that has references in it and disable this button if not, with a tooltip saying `This secret has no references` or something like that

File: frontend/src/views/SecretMainPage/components/SecretListView/SecretReferenceDetails.tsx
Comment (scott-ray-wilson) [inline frontend/src/views/SecretMainPage/components/SecretListView/SecretReferenceDetails.tsx:91]: I think children should be behind an expandable, so only the root references are shown initially giving the user the ability to expand each descendant instead of all being "expanded" initially

File: frontend/src/views/SecretMainPage/components/SecretListView/SecretReferenceDetails.tsx
Comment (scott-ray-wilson) [inline frontend/src/views/SecretMainPage/components/SecretListView/SecretReferenceDetails.tsx:119]: I don't think we should cap height here, I'd let it fully expand to dialog max (90% vh) and then just scroll the whole content. Also adding a little margin between the value and the children might look cleaner

File: frontend/src/views/SecretMainPage/components/SecretListView/SecretReferenceDetails.tsx
Comment (scott-ray-wilson) [inline frontend/src/views/SecretMainPage/components/SecretListView/SecretReferenceDetails.tsx:43]: I think this should be a little more clear that if you hover it shows the value - maybe an eye icon or even `Show Value` or `Reveal Value` - it kind of looks like a badge to me right now - not something that a user would intuitively click/hover over

File: backend/src/services/secret-v2-bridge/secret-v2-bridge-fns.ts
Comment (scott-ray-wilson) [inline backend/src/services/secret-v2-bridge/secret-v2-bridge-fns.ts:425]: Not necessarily saying to do so here but don't we need at some level to check permissions for each reference within the secret? Otherwise can't people potentially see a secret value they shouldn't via an expanded value/reference tree?

File: backend/src/services/secret-v2-bridge/secret-v2-bridge-service.ts
Comment (scott-ray-wilson) [inline backend/src/services/secret-v2-bridge/secret-v2-bridge-service.ts:1662]: Don't we need to check for permissions at each reference along the way in addition to the root secret?

File: backend/src/services/secret/secret-service.ts
Comment (scott-ray-wilson) [inline backend/src/services/secret/secret-service.ts:1084]: type `doesn't not` -> `does not`

File: frontend/src/views/SecretMainPage/components/SecretListView/SecretReferenceDetails.tsx
Comment (scott-ray-wilson) [inline frontend/src/views/SecretMainPage/components/SecretListView/SecretReferenceDetails.tsx:99]: could disabled query if secret contains no reference values (regex check)

**Review by scott-ray-wilson — CHANGES_REQUESTED** [review]
As a general comment on the PR - this new reference tree addition is awesome but I don't think users should have to navigate to this view to see the expanded secret value. 

This should be visible from the table itself (both overview and main pages), maybe a hover or button as mentioned [here](https://infisical-users.slack.com/archives/C06QSKDKU4S/p1727216814516129)

File: backend/package-lock.json
Comment (akhilmhdh) [reply backend/package-lock.json:1]: It is

File: backend/src/services/secret-v2-bridge/secret-v2-bridge-service.ts
Comment (akhilmhdh) [reply backend/src/services/secret-v2-bridge/secret-v2-bridge-service.ts:1708]: As i said in PR description. We are seperating it into so that stack trace is not executed on normal operation.  This is avoid the overhead of stack trace in secret normal operations

File: frontend/src/views/SecretMainPage/components/SecretListView/SecretDetailSidebar.tsx
Comment (akhilmhdh) [reply frontend/src/views/SecretMainPage/components/SecretListView/SecretDetailSidebar.tsx:432]: I don't think we really need to make it that much complex to be honest. Like its not like something compute heavy or anything like that.

File: frontend/src/views/SecretMainPage/components/SecretListView/SecretReferenceDetails.tsx
Comment (akhilmhdh) [reply frontend/src/views/SecretMainPage/components/SecretListView/SecretReferenceDetails.tsx:91]: I tried this but didn't like the experience. 

First most of references won't be in so many levels. So I don;t think there is much to expand. Second the space is already a bit smaller so i don't want to make interactions inside there

File: frontend/src/views/SecretMainPage/components/SecretListView/SecretReferenceDetails.tsx
Comment (akhilmhdh) [reply frontend/src/views/SecretMainPage/components/SecretListView/SecretReferenceDetails.tsx:119]: 90%VH modal is also bad. I think a max cap is fine to a good extend. May be ill increase some more but on a bigger screen 90%VH is awful experience

File: backend/src/services/secret-v2-bridge/secret-v2-bridge-service.ts
Comment (akhilmhdh) [reply backend/src/services/secret-v2-bridge/secret-v2-bridge-service.ts:1662]: https://github.com/Infisical/infisical/pull/2603/files/39013cbeee4e1d013a4e13bc68b4903ce28bf299#diff-7d02b63b99e49c50295fdd39e013c8e9294316950497e86f31e56be915dfa8caR460

File: frontend/src/views/SecretMainPage/components/SecretListView/SecretReferenceDetails.tsx
Comment (akhilmhdh) [reply frontend/src/views/SecretMainPage/components/SecretListView/SecretReferenceDetails.tsx:99]: I actually forgot the load condition, I'll add that. Its ok for not disabling on regex condition.

**Comment by akhilmhdh** [comment]
@scott-ray-wilson We don't want to show the entire tree on hover and next issue with hover is what about when they type, and all other cases it will popup. 

Second thing is i don't want to fill any more button on main page rows. It's already too much.

File: frontend/src/views/SecretMainPage/components/SecretListView/SecretReferenceDetails.tsx
Comment (akhilmhdh) [reply frontend/src/views/SecretMainPage/components/SecretListView/SecretReferenceDetails.tsx]: Changed to Reveal Value

File: backend/src/services/secret/secret-service.ts
Comment (akhilmhdh) [reply backend/src/services/secret/secret-service.ts]: Changed

File: backend/src/services/secret-v2-bridge/secret-v2-bridge-service.ts
Comment (scott-ray-wilson) [reply backend/src/services/secret-v2-bridge/secret-v2-bridge-service.ts:1708]: wouldn't passing a boolean to a single function solve this concern though instead of running it twice? you already have the boolean one layer down, why not just
```  
const expandSecret = async (inputSecret: {
    value?: string;
    skipMultilineEncoding?: boolean | null;
    secretPath: string;
    environment: string;
    shouldStackTrace?: boolean;
  }) => {
    if (!inputSecret.value) return inputSecret.value;

    const shouldExpand = Boolean(inputSecret.value?.match(INTERPOLATION_SYNTAX_REG));
    if (!shouldExpand) return inputSecret.value;

    const { expandedValue } = await recursivelyExpandSecret(inputSecret);

    return inputSecret.skipMultilineEncoding ? formatMultiValueEnv(expandedValue) : expandedValue;
  };```
  
  and only pass `shouldStackTrace` on new endpoint

Comment (akhilmhdh) [reply backend/src/services/secret-v2-bridge/secret-v2-bridge-service.ts:1708]: The return signature would be the issue. I would need to make an object. I don't want that when it's not for tracing.

Also its not doing so much work in the second run everything is already cached

Comment (scott-ray-wilson) [reply backend/src/services/secret-v2-bridge/secret-v2-bridge-service.ts:1708]: addressed

File: frontend/src/views/SecretMainPage/components/SecretListView/SecretDetailSidebar.tsx
Comment (scott-ray-wilson) [reply frontend/src/views/SecretMainPage/components/SecretListView/SecretDetailSidebar.tsx:432]: I just feel like it might be confusing to to see a reference tree button when no references are present?

Comment (akhilmhdh) [reply frontend/src/views/SecretMainPage/components/SecretListView/SecretDetailSidebar.tsx:432]: Ok then let's disable it

Comment (scott-ray-wilson) [reply frontend/src/views/SecretMainPage/components/SecretListView/SecretDetailSidebar.tsx:432]: or at least say no references in the dialog when it opens, just feels ambiguous

**Comment by akhilmhdh** [comment]
Update screenshot
Screenshot 2024-10-22 at 11 44 17 PM
Screenshot 2024-10-22 at 11 53 18 PM
Screenshot 2024-10-22 at 11 57 44 PM

**Review by scott-ray-wilson — APPROVED** [review]

**Pull request merged**
